metagraph/src/graph/representation/succinct/boss.cpp

```diff
@@ -2373,10 +2364,8 @@ void update_terminal_bits(
         set_bit(terminal->data(), path[i + max_length - 1], async);
     }
 
-    if (path.size() % max_length == 0) { // last node is terminal
-        callback(std::move(path), nullopt);
+    if (path.size() % max_length == 0) // last node is terminal
```

Comment:
Hmm, I think we should also set the previous nodes as near_terminal here.
Actually, I think the way I assigned near terminals in the entire method is not optimal. 
Every node on the path should have the near-terminal bit set, because it is by definition near a terminal. This way, if we later merge into a node on this path, we won't need to set a terminal node. I can make this change later if you add a todo, should be pretty easy.

Reply:
But if every node is a near terminal, this "near terminal" doesn't mean anything and will propagate forever, so there will be no anchors assigned. Right?

Reply:
What we can do though is start backward traversal from this position. And process all diff-paths that merge into this anchor. I implemented it for starting from sink nodes (see soon in the next PR).
That processes a large portion of the graph, but the loopy tricky parts are handled by this approach.
So what we can do is to invoke the same backward traversal thing from the newly created anchors.

Hm, actually then we don't need the near-terminal thing at all, it seems.
This thing was supposed to mark all nodes that merge into an anchor soon (say, this was path X of near-anchor nodes). So that if there is another path (call it Y) that merges somewhere into X, we don't create an anchor in the merge because we know that it's close to an anchor in X.

But, with the backward traversal thing, we could cover all such Y from X (if Y merges into X, Y will be completely covered by a backward traversal from X).

Will implement this in the next PR then.

Reply:
Yes, if we do backward traversal, near terminals are not needed.

Note that they don't propagate forever. A merge into a near terminal doesn't make the merge nodes near-terminal. So the max distance of propagation is still 2 * max_depth

Reply:
Maybe I misinterpreted your
> Every node on the path should have the near-terminal bit set, because it is by definition near a terminal.

On a different note, I realized that even with backward traversal running this with concurrently with multiple threads is tricky, so I decided to keep `near_terminal` for now.

Reply:
Right, more precisely I should say:
 > Every node on the path that is succeeded by a terminal node should have the near-terminal bit set, because it is by definition near a terminal.
So the only nodes that are *not* near terminal are the nodes between the last terminal node and the end of the path.